sklearn/feature_selection/_sequential.py

```diff
@@ -0,0 +1,200 @@
+        for more details.
+
+    Attributes
+    ----------
+    n_features_to_select_ : int
+        The number of features that were selected. It corresponds to
+        `n_features_to_select` unless the parameter was None.
+
+    support_ : ndarray of bool of shape (n_features,)
+        The mask of selected features.
+
+    Examples
+    --------
+    >>> from sklearn.feature_selection import SequentialFeatureSelector
+    >>> from sklearn.neighbors import KNeighborsClassifier
+    >>> from sklearn.datasets import load_iris
+    >>> X, y = load_iris(return_X_y=True)
+    >>> knn = KNeighborsClassifier(n_neighbors=3)
+    >>> sfs = SequentialFeatureSelector(knn, n_features_to_select=3)
+    >>> sfs.fit(X, y)
+    SequentialFeatureSelector(estimator=KNeighborsClassifier(n_neighbors=3),
+                              n_features_to_select=3)
+    >>> sfs.get_support()
+    array([ True, False,  True,  True])
+    >>> sfs.transform(X).shape
+    (150, 3)
+
+    See also
+    --------
+    RFE : Recursive feature elimination based on importance weights.
+    RFECV : Recursive feature elimination based on importance weights, with
+        automatic selection of the number of features.
+    SelectFromModel : Feature selection based on thresholds of importance
+        weights.
+    """
+    def __init__(self, estimator, *, n_features_to_select=None, forward=True,
+                 scoring=None, cv=5, n_jobs=None):
+
+        self.estimator = estimator
+        self.n_features_to_select = n_features_to_select
+        self.forward = forward
+        self.scoring = scoring
+        self.cv = cv
+        self.n_jobs = n_jobs
+
+    def fit(self, X, y):
+        """Learn the features to select.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Training vectors.
+        y : array-like of shape (n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : object
+        """
+
+        tags = self._get_tags()
+        X, y = self._validate_data(
+            X, y, accept_sparse="csc",
+            ensure_min_features=2,
+            force_all_finite=not tags.get('allow_nan', True),
+            multi_output=True
+        )
+
+        if self.n_features_to_select is None:
+            self.n_features_to_select_ = X.shape[1] // 2
+        else:
+            self.n_features_to_select_ = self.n_features_to_select
+
+        if not 1 <= self.n_features_to_select_ < X.shape[1]:
+            raise ValueError(
+                "n_features_to_select must be in [1, n_features - 1] = "
+                f"[1, {X.shape[1] - 1}]. Got {self.n_features_to_select_}."
+            )
+
+        cloned_estimator = clone(self.estimator)
+
+        # the current mask corresponds to the set of features:
+        # - that we have already *selected* if we do forward selection
+        # - that we have already *excluded* if we do backward selection
```

Comment:
Personally, I think having one function handle forward and back and the two semantics for `current_mask` will make this harder to maintain.

(I am comparing this to having two functions: one for forward and one for backwards, which would repeat some code, but will be much easier to follow down the line)

Reply:
If we were to have 2 functions, the only difference between would be these 2 lines:

```py

            if not self.forward:
                candidate_mask = ~candidate_mask 
```

and all of the rest would be common.

We can definitely split this into two separate functions in the future if we need to, but for now I think having a common handling for forward and backward is simpler and clearer.

Reply:
also note that "we should probably merge the forward and backward strategies" was a comment that was made about the original PR

Reply:
The differences are in `n_iterations`

```
n_iterations = (self.n_features_to_select_ if self.forward
                        else X.shape[1] - self.n_features_to_select_)
```

There is a comment in `_get_best_new_feature` which explains why we flip the mask:

```
			if not self.forward:
                # For backward selection, we transform candidate_mask into its
                # complement, i.e. we change its semantic from "features to
			    # remove" to "features to keep" because _safe_indexing only
        		# understands the latter
```

And comment on the semantics

```
        # the current mask corresponds to the set of features:
        # - that we have already *selected* if we do forward selection
        # - that we have already *excluded* if we do backward selection
```

There are big blocks of comments trying to walk a reader through the semantics. This feels like something that should be broken apart in code.

Anyways, I don't think its worth blocking. We can talk about splitting them up in another PR.

Reply:
> The differences are in n_iterations

This isn't related to the factorization, and this will need to be done even if we split into 2 functions

> There are big blocks of comments trying to walk a reader through the semantics

This is because, as you know, I think comments are important. I'm sure other devs would not have gone through the trouble of writing them, but I do.

I have pushed a version where the forward and backward pass are splitted into 2 different functions in https://github.com/scikit-learn/scikit-learn/pull/17857. Let me know what you think, but to me, the code screams "please factorize me"

Reply:
> This is because, as you know, I think comments are important. I'm sure other devs would not have gone through the trouble of writing them, but I do.

In this case, the comments are much needed because of the two semantics. I was thinking "if the mask didn't have these two semantics, then the comments would not need to reexplain the semantics".